### PR TITLE
break integration tests into seperate packages so that they run in parallel

### PR DESCRIPTION
**Pull request by mikedanese**
If this is broken then our tests aren't threadsafe.

ref #25940

### CODE REVIEW

**Comment by spxtr** [comment]
Many failures :(

**Comment by spxtr** [comment]
Is it reasonable to just use `go test`'s built-in parallelism? Then we can only use it for some tests but the others will remain serial.

**Comment by mikedanese** [comment]
@spxtr Yes that is possible and probably a good first pass. I'll give it a shot tomorrow.

**Comment by jsafrane** [comment]
Will it run in parallel if most tests do `integration.DeleteAllEtcdKeys()`? They will delete etcd keys of other tests running at the same time.

**Comment by mikedanese** [comment]
Hmmmm, this may not work. Would it be a bad idea to start an etcd per test suite?

**Comment by jsafrane** [comment]
see https://github.com/kubernetes/kubernetes/pull/27034 for start, some more official scripting around would be helpful

**Comment by wojtek-t** [comment]
@mikedanese - actuall yes, a common etcd would be a good thing to do. However, what we need to do then is to:
- change each test to use a dedicated namespace
- (potentially) delete this namespace at the end of the test

BTW - what we also should do is to move tests from cmd/integration and split them into single tests under test/integration. I already started looking into it in my free cycles.

**Comment by jsafrane** [comment]
My PV tests won't work even in namespaces - PersistentVolumes are not namespaced. I *could* put all storage tests into one big tests case, but again, it won't work in parallel with other instance of the same test running on the same etcd.

**Comment by spxtr** [comment]
Please reassign to me once this is ready for review.

**Comment by mikedanese** [comment]
> The remaining issue is that some tests are creating and deleting non-namespaced objects (PVs, Nodes) so those may not work when running in parallel.

Within a package, tests are still run serially. I ran all the integration tests 5 times and they didn't flake.

**Comment by lavalamp** [comment]
I did something a while ago that makes the framework use a new prefix in
etcd, so even things that make global objects shouldn't interfere with each
other any more.

On Wed, Jul 6, 2016 at 4:07 AM, Wojciech Tyczynski <notifications@github.com

**Comment by wojtek-t** [comment]
>
Within a package, tests are still run serially. I ran all the integration tests 5 times and they didn't flake.

OK - that seems enough for us (at least for now).

I will take a look later today. So far the gain is not enormous (it.e ~4m vs ~6m before), but as we will be adding more and more tests, this will have bigger impact.

**Comment by mikedanese** [comment]
```
ok  	k8s.io/kubernetes/test/integration/auth	2.669s
ok  	k8s.io/kubernetes/test/integration/client	7.791s
ok  	k8s.io/kubernetes/test/integration/configmap	0.220s
ok  	k8s.io/kubernetes/test/integration/garbagecollector	215.674s
ok  	k8s.io/kubernetes/test/integration/kubectl	0.865s
ok  	k8s.io/kubernetes/test/integration/master	2.137s
ok  	k8s.io/kubernetes/test/integration/metrics	0.345s
ok  	k8s.io/kubernetes/test/integration/openshift	0.102s
ok  	k8s.io/kubernetes/test/integration/persistentvolumes	36.591s
ok  	k8s.io/kubernetes/test/integration/pods	3.668s
ok  	k8s.io/kubernetes/test/integration/quota	0.540s
ok  	k8s.io/kubernetes/test/integration/scheduler	79.818s
ok  	k8s.io/kubernetes/test/integration/secrets	0.139s
ok  	k8s.io/kubernetes/test/integration/serviceaccount	23.447s
```

the overall time is dictated by the garbagecollector package which is taking 4 minutes. Some big gains will come from reducing the time those tests take.

**Comment by wojtek-t** [comment]
ohh - actually that's a good point; I've always though that scheduler_test contains the longest ones :)

File: test/integration/client/client_test.go [hunk 1]
```diff
@@ -16,7 +16,7 @@ See the License for the specific language governing permissions and
 limitations under the License.
 */
 
-package integration
+package client
```
Comment (wojtek-t) [inline test/integration/client/client_test.go]: I can't comment on the above file, but shouldn't package in rbac_test.go be changed to auth ?

File: test/integration/persistentvolumes/persistent_volumes_test.go [hunk 2]
```diff
@@ -27,8 +27,6 @@ import (
 	"testing"
 	"time"
 
-	"github.com/golang/glog"
```
Comment (wojtek-t) [inline test/integration/persistentvolumes/persistent_volumes_test.go]: The same comment for openshift_test.go - shouldn't the package be openshift there?

**Comment by wojtek-t** [comment]
2 comments - other than that lgtm

File: test/integration/client/client_test.go (hunk 1 above)
Comment (mikedanese) [inline test/integration/client/client_test.go]: Fixed.

File: test/integration/persistentvolumes/persistent_volumes_test.go (hunk 2 above)
Comment (mikedanese) [inline test/integration/persistentvolumes/persistent_volumes_test.go]: Fixed

**Pull request merged**